test/coinselector.spec.ts

```diff
@@ -0,0 +1,92 @@
+import { Coin } from '../src/wallet/coin';
+import { CoinSelector } from '../src/wallet/coin-selector.ts';
+import { Transaction } from 'bitcoinjs-lib';
+
+describe('CoinSelector', () => {
+    let coinSelector: CoinSelector;
+    const feeRate: number = 5;
+
+    beforeAll(() => {
+        coinSelector = new CoinSelector(feeRate);
+    });
+
+    it('should generate the correct cost of change', () => {
+        // Fixed value for the cost of change based on the current feeRate.
+        const correctCostOfChange = 495;
+
+        // Assert the correctness of the CoinSelector result.
+        expect(correctCostOfChange).toBe(coinSelector.costOfChange);
+    });
+
+    it.each([
+        {
+            testName: 'should add change to the fee if it is dust',
+            testCoinValues: [
+                1005, 9040, 6440, 2340, 7540, 3920, 5705, 9030, 1092, 5009,
+            ],
+            expectedChange: 0, // Change will be zero when less than dust.
+            expectedCoinCount: 10, // All the coins are going to be selected.
+        },
+        {
+            testName: 'change should be considered when greater than dust',
+            testCoinValues: [5000, 4000],
+            txOutValue: 1000, // Less output so that the change is greater than dust.
+            expectedChange: 2185, // expectedChange = totalSelectedCoin - (txOutValue + transactionFees).
+            expectedCoinCount: 1, // One Coin will suffice for a 1000 sats output for this case.
+        },
+    ])(
+        '%s',
+        ({ txOutValue, testCoinValues, expectedChange, expectedCoinCount }) => {
+            const coins: Coin[] = [];
+            let totalBalance: number = 0;
+
+            // Calculate the total balance based on the fixed array of coin values.
+            for (const value of testCoinValues) {
+                const coin = new Coin({ value });
+                coins.push(coin);
+                // Calculate the total balance by subtracting the spending fee.
+                totalBalance += coin.value - coin.estimateSpendingFee(feeRate);
+            }
+            
+            // Create a new Transaction
+            const transaction: Transaction = new Transaction();
+
+            // Fixed fees for the transaction calculated based on the total size of the transaction and the change output size.
+            // transactionFees = (virtualSize + changeOutputSize) * feeRate, P2WPKH output is 31 bytes.
+            const transactionFees = 250;
```

Comment:
how is the fee for both transactions the same? Both transactions have different number of inputs, so fee for both of them should be different.
Something is wrong here.

Reply:
I don't think so, `transactionFees` account for these calculations in the coin-selector. Here, the inputs have not been added to the transaction yet. If you do a `console.log()` for the fees calculations in lines `26 to 28`, it gives the same value for both tests. My assumption was that it only considers the number of outputs, and there is only one in both tests.

https://github.com/Bitshala-Incubator/silent-pay/blob/main/packages/wallet/src/coin-selector.ts#L26

Reply:
Anyway, I've removed the `transactionFees` and set static values for `txOutValue` in both tests. Therefore, we don't have to deal with transaction fees. Additionally, since all coins are selected in both tests, this approach won't pose any problems in the future, even when more coin selection algorithms are implemented.